proposals/td_for_tdir.md

```diff
@@ -18,15 +18,22 @@ Over time the team [realised](https://github.com/mozilla-iot/schemas/issues/35)
 ### Example
 
 Example Thing Description for a gateway:
-```
+```json
 {
   "title": "WebThings Gateway",
   "@context": "https://iot.mozilla.org/schemas/",
   "@type": "Gateway",
   "actions": {
+    "startPairing": {
+      "name": "Start Pairing",
+      "@type": "StartPairingAction"
+    },
     "addThing": {
       "title": "Add Device",
       "@type": "AddThingAction",
+      "input": {
+        "type": "object"
+      }
```

Comment:
To be honest I'm not really sure how this action would work. What would the input actually be? It feels wrong for it to just be an object with undefined contents. Could you pass it the URL of a Thing Description to add to the directory perhaps? (Note this could be a cloud service acting as a directory as opposed to a gateway).

My (unwritten) assumption was that the string argument to "RemoveThingAction" was a Thing Description URL or id. I would therefore suggest this should be a string as well.

Reply:
Right now, we end up sending the entire thing description from the UI to the gateway. It'd be nice if the entire TD model were available as a type.

For the remove action, yes, we just use the ID.

Reply:
> Right now, we end up sending the entire thing description from the UI to the gateway. It'd be nice if the entire TD model were available as a type.

Hmm. For the purposes of standardisation for generic directories, I'd suggest it would be better to send the URL of the Thing Description and have the directory keep a reference to it. The problem with just sending the contents of the Thing Description is that it assumes the Thing Description is a static resource that never changes. By adding a thing by its Thing Description URL, the directory can keep a reference to that URL and fetch new versions of the Thing Description as they are updated, or in some cases serve it directly from the source. Some directories may just link to Thing Descriptions rather than re-publish them at another URL like the WebThings Gateway does.

Reply:
I don't disagree, but in the case of the gateway, you can't just send the URL because it doesn't actually _exist_ on the gateway yet.

Reply:
Ah, maybe we're thinking of different use cases. In what circumstances does the front end send the Thing Description? I'm thinking of the case where you add a thing by its URL in the UI.

Reply:
Even in that case, we still end up sending the entire thing description to the gateway when the thing is "saved".

Reply:
I'm less concerned about what the WebThings Gateway currently does, and more concerned about what a generic directory should do. Sending the contents of the Thing Description seems like the wrong thing to do, the directory at least needs to know the original URL in order to fetch updates. If you're going to send both the contents of the Thing Description and the URL, you may as well just send the URL and have the directory fetch the Thing Description itself (if it even needs to). Unless I'm missing something?

If there's a use case where the thing doesn't yet have a URL then it's by definition not a web thing that's being added and maybe that should be a separate action?

Reply:
I think we're thinking of two separate use cases here:
1. Adding the contents of a Thing Description to a database to create a new web thing that didn't exist before, in the case of a gateway which is responsible for bridging or proxying a device from another protocol or origin and exposing the Web Thing API itself
2. Adding an existing web thing with an existing URL to a directory of web things

My assumption is that the `AddThingAction` is the latter, on the basis that not all directories are gateways.

Reply:
Ok, that's a fair point. Maybe this is another separate gateway-specific action then, to go along with StartPairingAction. Should we add a note about that as well?

Reply:
It may be something that's quite specific to Mozilla's particular implementation (there are probably other ways of implementing the same thing), but if you can think of a way to word it which is widely applicable then please feel free. Otherwise we could just leave it out.

So I'd suggest making the input of the `AddThingAction` a string. Thank you for figuring this out with me!

Reply:
> So I'd suggest making the input of the `AddThingAction` a string. Thank you for figuring this out with me!

Already done.